src/file/create.h

```diff
@@ -177,47 +109,43 @@ static __always_inline void exit_create(void *ctx, file_link_type_t link_type)
         // dentry we captured during the security path may not be
         // assigned an inode during hardlinks since the real dentry
         // exists elsewhere with the same inode as the source
-        ret = file_from_dentry(event.source, &fm->u.action.target, &fm->u.action.target_owner);
+        ret = file_from_dentry(event->create.source, &fm->u.action.target, &fm->u.action.target_owner);
     } else {
         // `inode` could be NULL for kernel pseudo filesystems such as
         // cgroupfs. We don't care about these filesystems so skip if
         // inode is NULL. There may be false negatives here (inodes
         // being NULL for other reasons) but in our testing we have
         // only ever triggered this from cgroupfs
-        ret = read_field(event.target_dentry, CRC_DENTRY_D_INODE, &inode, sizeof(inode));
+        ret = read_field(event->target_dentry, CRC_DENTRY_D_INODE, &inode, sizeof(inode));
         if (ret < 0) goto EmitWarning;
         if (inode == NULL) goto NoEvent;
         ret = file_and_owner_from_ino(inode, &fm->u.action.target, &fm->u.action.target_owner);
     }
 
     if (ret < 0) goto EmitWarning;
 
-    fm->type = FM_CREATE;
-    fm->u.action.pid = event.pid_tgid >> 32;
-    fm->u.action.mono_ns = event.start_ktime_ns;
+    fm->u.action.pid = pid_tgid >> 32;
+    fm->u.action.mono_ns = event->start_ktime_ns;
     fm->u.action.buffer_len = sizeof(file_message_t);
     fm->u.action.u.create.source_link = link_type;
 
-    init_filtered_cached_path(cached_path, event.target_dentry, event.target_vfsmount);
+    init_filtered_cached_path(cached_path, event->target_dentry, event->vfsmount);
     if (link_type == LINK_HARD) {
-        cached_path->next_path = event.source;
+        cached_path->next_path = event->create.source;
     }
 
-    bpf_tail_call(ctx, &tp_programs, FILE_PATHS);
+    return fm;
 
-    info.tailcall = FILE_PATHS;
-    set_local_warning(W_TAIL_CALL_MAX, info);
-    goto EmitWarning;
-
- EventMismatch:
-    info.stored_pid_tgid = event.pid_tgid;
-    set_local_warning(W_PID_TGID_MISMATCH, info);
+ NoEvent:
+    fm = NULL;
+    goto Done;
 
  EmitWarning:
-    push_file_warning(ctx, fm, FM_CREATE);
+    fm->u.warning.message_type.file = fm->type;
+    fm->type = FM_WARNING;
 
- NoEvent:
-    // lookup tail calls completed; ensure we re-init cached_path next call
+ Done:
+    // reset next_dentry for this CPU
     cached_path->next_dentry = NULL;
-    return;
+    return NULL;
```

Comment:
question: How come we return `NULL` on `Done` unlike the other exit file types where we return a file message? Will this negate emitting the warning?

Reply:
Because that was a bug, good catch! We need to return `fm` here to then emit the warning.